File: projects/kotlin/src/main/kotlin/StaterStateMachine.kt
package org.example

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue


interface Context

interface ContextJsonAdapter<C : Context> {
    fun toJson(context: C): String
    fun fromJson(json: String): C
}

typealias Event<C> = (context: C) -> Unit
typealias NameEvent<C> = (name: String, context: C) -> Unit
typealias StateEvent<T, C> = (state: T, context: C) -> Unit
typealias TransitionMiddleware<C> = (C, Event<C>) -> Unit
typealias TransitionNameMiddleware<C> = (String, C, NameEvent<C>) -> Unit
typealias StateMachineFactory<T, C> = (
    transitions: List<Transition<T, C>>,
    state: T,
    states: List<T>,
    context: C,
    transitionMiddlewares: Map<String, List<TransitionMiddleware<C>>>,
    transitionAllMiddlewares: List<TransitionNameMiddleware<C>>,
    transitionCallbacks: Map<String, List<Event<C>>>,
    transitionAllCallbacks: List<NameEvent<C>>,
    stateCallbacks: Map<T, List<Event<C>>>,
    stateAllCallbacks: List<StateEvent<T, C>>,
    contextJsonAdapter: ContextJsonAdapter<C>?,
) -> StaterStateMachine<T, C>

data class Transition<T, C : Context>(
    val name: String,
    val start: T,
    val end: T,
    @JsonIgnore val condition: ((C) -> Boolean)? = null,
    @JsonIgnore val event: Event<C>? = null,
)

data class JsonSchema<T>(
    val states: List<T>,
    val startState: T,
    val transitions: List<Transition<T, *>>,
)

data class JsonState<T>(
    val state: T,
    val context: String,
)

abstract class StaterStateMachine<T, C : Context>(
    private val transitions: List<Transition<T, C>>,
    private var startState: T,
    private val states: List<T>,
    private var context: C,
    private var state: T = startState,
    private val transitionsGroupedStart: Map<T, List<Transition<T, C>>> = transitions.groupBy { it.start },
    private val transitionsByName: Map<String, Transition<T, C>> = transitions.associateBy { it.name },

    private val transitionMiddlewares: Map<String, List<TransitionMiddleware<C>>> = mapOf(),
    private val transitionAllMiddlewares: List<TransitionNameMiddleware<C>> = listOf(),

    private val transitionCallbacks: Map<String, List<Event<C>>> = mapOf(),
    private val transitionAllCallbacks: List<NameEvent<C>> = listOf(),

    private val stateCallbacks: Map<T, List<Event<C>>> = mapOf(),
    private val stateAllCallbacks: List<StateEvent<T, C>> = listOf(),
    private var contextJsonAdapter: ContextJsonAdapter<C>? = null,
) {
    fun getState(): T = state
    fun getContext(): C = context

    fun transition(name: String) {
        val transition = transitionsByName[name] ?: error("Transition not found: $name")

        fun conditionHandler() {
            if (state != transition.start) {
                error("Start state does not match transition's start state: ${transition.start}")
            }
            transition.condition?.let {
                if (!it(context)) {
                    error("Condition return false for transition $name")
                }
            }
        }

        var index = 0
        var index2 = 0
        val transitionMiddleware = transitionMiddlewares[name]

        fun internalNext(context: C) {
            return if (transitionMiddleware != null && index2 < transitionMiddleware.size) {
                transitionMiddleware[index2++](context, ::internalNext)
            } else {
                conditionHandler()
            }
        }

        fun next(name: String, context: C) {
            return if (index < transitionAllMiddlewares.size) {
                transitionAllMiddlewares[index++](name, context, ::next)
            } else {
                internalNext(context)
            }
        }

        next(name, context)

        state = transition.end
        transition.event?.let { it(context) }

        transitionAllCallbacks.forEach { it(name, context) }
        transitionCallbacks[name]?.forEach { it(context) }
        stateAllCallbacks.forEach { it(state, context) }
        stateCallbacks[state]?.forEach { it(context) }
    }

    fun autoTransition() {
        (transitionsGroupedStart[state] ?: emptyList()).forEach {
            try {
                transition(it.name)
                return@forEach
            } catch (_: Exception) {
            }
        }
    }

    fun toJsonSchema() = jacksonObjectMapper().writeValueAsString(
        JsonSchema(
            states = states, startState = startState, transitions = transitions
        )
    ) ?: error("Cant serialize")

    fun toJson() = jacksonObjectMapper().writeValueAsString(
        JsonState(
            state = state, context = contextJsonAdapter!!.toJson(context)
        )
    ) ?: error("Cant serialize")

    fun fromJson(json: String, stateConverter: (String) -> T) =
        jacksonObjectMapper().readValue<JsonState<T>>(json).let {
            this.state = stateConverter(it.state.toString())
            this.context = contextJsonAdapter!!.fromJson(it.context)
        }
}


class BaseFSM<T, C : Context>(
    transitions: List<Transition<T, C>>,
    startState: T,
    states: List<T>,
    context: C,
    transitionMiddlewares: Map<String, List<TransitionMiddleware<C>>>,
    transitionAllMiddlewares: List<TransitionNameMiddleware<C>>,
    transitionCallbacks: Map<String, List<Event<C>>>,
    transitionAllCallbacks: List<NameEvent<C>>,
    stateCallbacks: Map<T, List<Event<C>>>,
    stateAllCallbacks: List<StateEvent<T, C>>,
    contextJsonAdapter: ContextJsonAdapter<C>?,
) : StaterStateMachine<T, C>(
    transitions = transitions,
    startState = startState,
    states = states,
    context = context,
    transitionMiddlewares = transitionMiddlewares,
    transitionAllMiddlewares = transitionAllMiddlewares,
    transitionCallbacks = transitionCallbacks,
    transitionAllCallbacks = transitionAllCallbacks,
    stateCallbacks = stateCallbacks,
    stateAllCallbacks = stateAllCallbacks,
    contextJsonAdapter = contextJsonAdapter
)


class StaterStateMachineBuilder<T, C : Context>(
    private val transitions: MutableMap<String, Transition<T, C>> = mutableMapOf(),
    private var state: T? = null,
    private var states: MutableList<T> = mutableListOf(),
    private var context: C? = null,

    private val transitionMiddlewares: MutableMap<String, MutableList<TransitionMiddleware<C>>> = mutableMapOf(),
    private val transitionAllMiddlewares: MutableList<TransitionNameMiddleware<C>> = mutableListOf(),

    private val transitionCallbacks: MutableMap<String, MutableList<Event<C>>> = mutableMapOf(),
    private val transitionAllCallbacks: MutableList<NameEvent<C>> = mutableListOf(),

    private val stateCallbacks: MutableMap<T, MutableList<Event<C>>> = mutableMapOf(),
    private val stateAllCallbacks: MutableList<StateEvent<T, C>> = mutableListOf(),
    private var factory: StateMachineFactory<T, C> = {
            transitionsA,
            startStateA,
            statesA,
            contextA,
            transitionMiddlewaresA,
            transitionAllMiddlewaresA,
            transitionCallbacksA,
            transitionAllCallbacksA,
            stateCallbacksA,
            stateAllCallbacksA,
            contextJsonAdapterA,
        ->
        BaseFSM(
            transitions = transitionsA,
            startState = startStateA,
            states = statesA,
            context = contextA,
            transitionMiddlewares = transitionMiddlewaresA,
            transitionAllMiddlewares = transitionAllMiddlewaresA,
            transitionCallbacks = transitionCallbacksA,
            transitionAllCallbacks = transitionAllCallbacksA,
            stateCallbacks = stateCallbacksA,
            stateAllCallbacks = stateAllCallbacksA,
            contextJsonAdapter = contextJsonAdapterA
        )
    },
    private var contextJsonAdapter: ContextJsonAdapter<C>? = null,
) {
    fun addTransition(
        name: String,
        start: T,
        end: T,
        condition: ((C) -> Boolean)? = null,
        event: ((C) -> Unit)? = null,
    ) = apply {
        addState(start)
        addState(end)
        transitions[name] = Transition(name, start, end, condition, event)
    }

    fun addState(state: T) = apply {
        if (!states.contains(state)) {
            states.add(state)
        }
    }

    fun setTransitionCondition(name: String, condition: ((C) -> Boolean)) = apply {
        val transition = transitions[name] ?: error("Transition not found: $name")
        transitions[name] = transition.copy(condition = condition)
    }

    fun setTransitionEvent(name: String, event: ((C) -> Unit)) = apply {
        val transition = transitions[name] ?: error("Transition not found: $name")
        transitions[name] = transition.copy(event = event)
    }

    fun transitionMiddleware(name: String, middleware: TransitionMiddleware<C>) = apply {
        val middlewares = transitionMiddlewares.getOrDefault(name, mutableListOf())
        middlewares.add(middleware)
        transitionMiddlewares[name] = middlewares
    }

    fun transitionAllMiddleware(middleware: TransitionNameMiddleware<C>) = apply {
        transitionAllMiddlewares.add(middleware)
    }

    fun subscribeOnTransition(name: String, callback: Event<C>) = apply {
        val callbacks = transitionCallbacks.getOrDefault(name, mutableListOf())
        callbacks.add(callback)
        transitionCallbacks[name] = callbacks
    }

    fun subscribeOnAllTransition(callback: NameEvent<C>) = apply { transitionAllCallbacks.add(callback) }

    fun subscribeOnState(state: T, callback: Event<C>) = apply {
        val callbacks = stateCallbacks.getOrDefault(state, mutableListOf())
        callbacks.add(callback)
        stateCallbacks[state] = callbacks
    }

    fun subscribeOnAllState(callback: StateEvent<T, C>) = apply { stateAllCallbacks.add(callback) }

    fun setStartState(state: T) = apply { this.state = state }
    fun setContext(context: C) = apply { this.context = context }

    fun fromJsonSchema(schema: String, stateConverter: (String) -> T) = apply {
        jacksonObjectMapper().readValue<JsonSchema<T>>(schema).let {
            it.states.forEach { it2 -> addState(stateConverter(it2.toString())) }
            it.transitions.forEach { it2 ->
                addTransition(
                    it2.name,
                    stateConverter(it2.start.toString()),
                    stateConverter(it2.end.toString())
                )
            }
            setStartState(stateConverter(it.startState.toString()))
        }
    }

    fun setFactory(factory: StateMachineFactory<T, C>) = apply { this.factory = factory }

    fun setContextJsonAdapter(contextJsonAdapter: ContextJsonAdapter<C>) = apply {
        this.contextJsonAdapter = contextJsonAdapter
    }

    fun build(): StaterStateMachine<T, C> {
        return factory(
            transitions.values.toList(),
            state ?: error("Start state must be set"),
            states,
            context ?: error("Context must be set"),
            transitionMiddlewares,
            transitionAllMiddlewares,
            transitionCallbacks,
            transitionAllCallbacks,
            stateCallbacks,
            stateAllCallbacks,
            contextJsonAdapter
        )
    }
}

File: projects/ts/src/StaterStateMachine.ts
export interface Context {
}

export interface ContextJsonAdapter<C> {
    toJson(context: C): string

    fromJson(json: string): C
}

export type FSMEvent<C> = (context: C) => void
export type NameFSMEvent<C> = (name: string, context: C) => void
export type StateFSMEvent<T, C> = (state: T, context: C) => void
export type TransitionMiddleware<C> = (context: C, FSMEvent: FSMEvent<C>) => void
export type TransitionNameMiddleware<C> = (name: string, context: C, FSMEvent: NameFSMEvent<C>) => void
export type StateMachineFactory<T, C extends Context> = (
    transitions: Transition<T, C>[],
    state: T,
    states: T[],
    context: C,
    transitionMiddlewares: Map<string, TransitionMiddleware<C>[]>,
    transitionAllMiddlewares: TransitionNameMiddleware<C>[],
    transitionCallbacks: Map<string, FSMEvent<C>[]>,
    transitionAllCallbacks: NameFSMEvent<C>[],
    stateCallbacks: Map<T, FSMEvent<C>[]>,
    stateAllCallbacks: StateFSMEvent<T, C>[],
    contextJsonAdapter?: ContextJsonAdapter<C>
) => StaterStateMachine<T, C>;

export interface Transition<T, C extends Context> {
    name: string
    start: T
    end: T
    condition?: ((context: C) => boolean)
    event?: FSMEvent<C>
}

interface JsonSchema<T> {
    states: T[]
    startState: T
    transitions: Transition<T, Context>[]
}

interface JsonState<T> {
    state: T
    context: string
}


export abstract class StaterStateMachine<T, C extends Context> {
    private state: T;
    private transitionsGroupedStart: Map<T, Transition<T, C>[]>;
    private transitionsByName: Map<string, Transition<T, C>>;
    private transitionMiddlewares: Map<string, TransitionMiddleware<C>[]>;
    private transitionAllMiddlewares: TransitionNameMiddleware<C>[];
    private transitionCallbacks: Map<string, FSMEvent<C>[]>;
    private transitionAllCallbacks: NameFSMEvent<C>[];
    private stateCallbacks: Map<T, FSMEvent<C>[]>;
    private stateAllCallbacks: StateFSMEvent<T, C>[];
    private contextJsonAdapter?: ContextJsonAdapter<C>;

    constructor(
        private transitions: Transition<T, C>[],
        private startState: T,
        private states: T[],
        private context: C,
        transitionMiddlewares: Map<string, TransitionMiddleware<C>[]> = new Map(),
        transitionAllMiddlewares: TransitionNameMiddleware<C>[] = [],
        transitionCallbacks: Map<string, FSMEvent<C>[]> = new Map(),
        transitionAllCallbacks: NameFSMEvent<C>[] = [],
        stateCallbacks: Map<T, FSMEvent<C>[]> = new Map(),
        stateAllCallbacks: StateFSMEvent<T, C>[] = [],
        contextJsonAdapter?: ContextJsonAdapter<C>,
        state: T = startState
    ) {
        this.state = state;
        this.transitionsGroupedStart = new Map();
        this.transitionsByName = new Map();
        transitions.forEach(el => {
            this.transitionsByName.set(el.name, el)
            let grouped = this.transitionsGroupedStart.get(el.start)
            if (grouped) {
                grouped.push(el)
            } else {
                this.transitionsGroupedStart.set(el.start, [el])
            }
        })
        this.transitionMiddlewares = transitionMiddlewares;
        this.transitionAllMiddlewares = transitionAllMiddlewares;
        this.transitionCallbacks = transitionCallbacks;
        this.transitionAllCallbacks = transitionAllCallbacks;
        this.stateCallbacks = stateCallbacks;
        this.stateAllCallbacks = stateAllCallbacks;
        this.contextJsonAdapter = contextJsonAdapter;
    }

    public getState(): T {
        return this.state;
    }

    public getContext(): C {
        return this.context;
    }

    public transition(name: string): void {
        const transition = this.transitionsByName.get(name);
        if (!transition) {
            throw new Error(`Transition not found: ${name}`);
        }

        const conditionHandler = () => {
            if (this.state !== transition.start) {
                throw new Error(`Start state does not match transition's start state: ${transition.start}`);
            }
            if (transition.condition && !transition.condition(this.context)) {
                throw new Error(`Condition return false for transition ${name}`);
            }
        };

        let index = 0;
        let index2 = 0;
        const transitionMiddleware = this.transitionMiddlewares.get(name);

        const internalNext = (context: C): void => {
            if (transitionMiddleware && index2 < transitionMiddleware.length) {
                transitionMiddleware[index2++](context, internalNext);
            } else {
                conditionHandler();
            }
        };

        const next = (name: string, context: C): void => {
            if (index < this.transitionAllMiddlewares.length) {
                this.transitionAllMiddlewares[index++](name, context, next);
            } else {
                internalNext(context);
            }
        };

        next(name, this.context);

        this.state = transition.end;
        if (transition.event) {
            transition.event(this.context);
        }

        this.transitionAllCallbacks.forEach(callback => callback(name, this.context));
        this.transitionCallbacks.get(name)?.forEach(callback => callback(this.context));
        this.stateAllCallbacks.forEach(callback => callback(this.state, this.context));
        this.stateCallbacks.get(this.state)?.forEach(callback => callback(this.context));
    }

    public autoTransition(): void {
        const transitionsForState = this.transitionsGroupedStart.get(this.state) || [];
        for (const transition of transitionsForState) {
            try {
                this.transition(transition.name);
                return;
            } catch (_) {
            }
        }
    }

    public toJsonSchema(): string {
        return JSON.stringify({
            states: this.states,
            startState: this.startState,
            transitions: this.transitions
        });
    }

    public toJson(): string {
        if (!this.contextJsonAdapter) {
            throw new Error('ContextJsonAdapter is not provided');
        }
        return JSON.stringify({
            state: this.state,
            context: this.contextJsonAdapter.toJson(this.context)
        });
    }

    public fromJson(json: string, stateConverter: (state: string) => T): void {
        const jsonState = JSON.parse(json) as JsonState<T>;
        this.state = stateConverter(String(jsonState.state));
        if (this.contextJsonAdapter) {
            this.context = this.contextJsonAdapter.fromJson(jsonState.context);
        }
    }
}


class BaseFSM<T, C extends Context> extends StaterStateMachine<T, C> {
    constructor(
        transitions: Transition<T, C>[],
        startState: T,
        states: T[],
        context: C,
        transitionMiddlewares: Map<string, TransitionMiddleware<C>[]>,
        transitionAllMiddlewares: TransitionNameMiddleware<C>[],
        transitionCallbacks: Map<string, FSMEvent<C>[]>,
        transitionAllCallbacks: NameFSMEvent<C>[],
        stateCallbacks: Map<T, FSMEvent<C>[]>,
        stateAllCallbacks: StateFSMEvent<T, C>[],
        contextJsonAdapter?: ContextJsonAdapter<C>
    ) {
        super(
            transitions,
            startState,
            states,
            context,
            transitionMiddlewares,
            transitionAllMiddlewares,
            transitionCallbacks,
            transitionAllCallbacks,
            stateCallbacks,
            stateAllCallbacks,
            contextJsonAdapter
        );
    }
}


export class StaterStateMachineBuilder<T extends string, C extends Context> {
    private transitions: Map<string, Transition<T, C>> = new Map();
    private state: T | null = null;
    private states: T[] = [];
    private context: C | null = null;

    private transitionMiddlewares: Map<string, TransitionMiddleware<C>[]> = new Map();
    private transitionAllMiddlewares: TransitionNameMiddleware<C>[] = [];

    private transitionCallbacks: Map<string, FSMEvent<C>[]> = new Map();
    private transitionAllCallbacks: NameFSMEvent<C>[] = [];

    private stateCallbacks: Map<T, FSMEvent<C>[]> = new Map();
    private stateAllCallbacks: StateFSMEvent<T, C>[] = [];

    private factory: StateMachineFactory<T, C> = (
        transitions,
        startState,
        states,
        context,
        transitionMiddlewares,
        transitionAllMiddlewares,
        transitionCallbacks,
        transitionAllCallbacks,
        stateCallbacks,
        stateAllCallbacks,
        contextJsonAdapter
    ) => {
        return new BaseFSM(
            transitions,
            startState,
            states,
            context,
            transitionMiddlewares,
            transitionAllMiddlewares,
            transitionCallbacks,
            transitionAllCallbacks,
            stateCallbacks,
            stateAllCallbacks,
            contextJsonAdapter
        );
    };

    private contextJsonAdapter?: ContextJsonAdapter<C>;

    addTransition(
        name: string,
        start: T,
        end: T,
        condition?: (context: C) => boolean,
        event?: FSMEvent<C>
    ): this {
        this.addState(start);
        this.addState(end);
        this.transitions.set(name, {name, start, end, condition, event});
        return this;
    }

    addState(state: T): this {
        if (!this.states.includes(state)) {
            this.states.push(state);
        }
        return this;
    }

    setTransitionCondition(name: string, condition: (context: C) => boolean): this {
        const transition = this.transitions.get(name);
        if (!transition) throw new Error(`Transition not found: ${name}`);
        transition.condition = condition;
        return this;
    }

    setTransitionEvent(name: string, event: (context: C) => void): this {
        const transition = this.transitions.get(name);
        if (!transition) throw new Error(`Transition not found: ${name}`);
        transition.event = event;
        return this;
    }

    transitionMiddleware(name: string, middleware: TransitionMiddleware<C>): this {
        const middlewares = this.transitionMiddlewares.get(name) || [];
        middlewares.push(middleware);
        this.transitionMiddlewares.set(name, middlewares);
        return this;
    }

    transitionAllMiddleware(middleware: TransitionNameMiddleware<C>): this {
        this.transitionAllMiddlewares.push(middleware);
        return this;
    }

    subscribeOnTransition(name: string, callback: FSMEvent<C>): this {
        const callbacks = this.transitionCallbacks.get(name) || [];
        callbacks.push(callback);
        this.transitionCallbacks.set(name, callbacks);
        return this;
    }

    subscribeOnAllTransition(callback: NameFSMEvent<C>): this {
        this.transitionAllCallbacks.push(callback);
        return this;
    }

    subscribeOnState(state: T, callback: FSMEvent<C>): this {
        const callbacks = this.stateCallbacks.get(state) || [];
        callbacks.push(callback);
        this.stateCallbacks.set(state, callbacks);
        return this;
    }

    subscribeOnAllState(callback: StateFSMEvent<T, C>): this {
        this.stateAllCallbacks.push(callback);
        return this;
    }

    setStartState(state: T): this {
        this.state = state;
        return this;
    }

    setContext(context: C): this {
        this.context = context;
        return this;
    }

    fromJsonSchema(schema: string, stateConverter: (state: string) => T): this {
        const jsonSchema: JsonSchema<T> = JSON.parse(schema);
        jsonSchema.states.forEach((state: string) => this.addState(stateConverter(state)));
        jsonSchema.transitions.forEach((transition) => {
            this.addTransition(
                transition.name,
                stateConverter(transition.start),
                stateConverter(transition.end)
            );
        });
        this.setStartState(stateConverter(jsonSchema.startState));
        return this;
    }

    setFactory(factory: StateMachineFactory<T, C>): this {
        this.factory = factory;
        return this;
    }

    setContextJsonAdapter(contextJsonAdapter: ContextJsonAdapter<C>): this {
        this.contextJsonAdapter = contextJsonAdapter;
        return this;
    }

    build(): StaterStateMachine<T, C> {
        if (!this.state || !this.context) {
            throw new Error("Start state and context must be set");
        }

        return this.factory(
            Array.from(this.transitions.values()),
            this.state,
            this.states,
            this.context,
            this.transitionMiddlewares,
            this.transitionAllMiddlewares,
            this.transitionCallbacks,
            this.transitionAllCallbacks,
            this.stateCallbacks,
            this.stateAllCallbacks,
            this.contextJsonAdapter
        );
    }
}